### PR TITLE
[WIP] Typescript and Screeps, from Beginner to Advanced Contrib Article

### PR DESCRIPTION
**Pull request by bryanbecker**
This PR is currently a work in a progress.  I will remove the tag when completed.

The guide will include the following content:

- Setting up a bare project with typescript
- Installing typings
- _Brief_ intro to third party utils (such as typescript starter, typings, webpack) and integrating them into the project
- Advanced usage of the typescript compiler to for working with Screeps objects.  Including: how to type the Memory object, decorators, typeguards and generics for Screeps objects

### CODE REVIEW

**Comment by bryanbecker** [comment]
[WIP] tag removed

**Comment by bryanbecker** [comment]
@artch 

I added a clarification to my 2nd regarding the limitations of type guards.

I have a proposed and working solution with Type Tags I'd like to write about, but the article was already getting quite long.

I would like to write a final forth article covering this solution, along with other "Tips & Tricks"... A collection of a few interesting TypeScript / Screeps tricks that are too small to warrant their own article.

But I'm reluctant to write article 4 until I get some feedback on these three.  Any thoughts?

**Comment by artch** [comment]
This PR has a lot of content. Since we have no TypeScript experts in our team, it'd be nice to wait for some feedback from the community.

Also, I'm not sure whether this subject should contain so much content and explanations divided into 3 articles, or rather be written in a more concise how-to manner.

File: source/contributed/typescript-and-screeps-1.md [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+
+Part One of this series with introduce setting up a TypeScript project for Screeps.  This article will present a basic guide to setting up a TypeScript project for Screeps, and a brief introduction to third party tools.
+
+For users who already have a working TypeScript setup, you can directly skip to [Part 2](http://docs.screeps.com/contributed/typescript-and-screeps-2.html).
+
+For a quick-start, skip down to [screeps-typescript-starter](#screeps-typescript-starter).
+
+## Why TypeScript?
+
+TypeScript is a superset of ECMAScript (a fancy way to say Javascript).  That means it has *all* the features of ECMA, *plus* some extra ones.  When a project is compiled, Typescript will process all your files, and "transpile" them into the target Javascript version (ES6, in the case of Screeps).  This works similarly to Babel, if you have used that in the past.
```
Comment (PostCrafter) [inline source/contributed/typescript-and-screeps-1.md]: > (a fancy way to say Javascript)
Not quite true, ECMAScript is a specification that JavaScript implements, even though ECMAScript was created to standardize JavaScript, there can be (and are) other implementations.

File: source/contributed/typescript-and-screeps-1.md [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+
+Using TypeScript can be frustrating when you first start, because it seems the compiler never stops complaining. Power through it. This ends up being the most useful feature of TypeScript once you get going. Towards then, we'll look at some concrete examples of this.
+
+This guide won't attempt to teach you typescript ([Typescript Deep Dive](https://basarat.gitbooks.io/typescript/docs/why-typescript.html) is an execellent starting resource for that), but will provide an introduction in getting set up, and using the more advanced features to your advantage
```
Comment (PostCrafter) [inline source/contributed/typescript-and-screeps-1.md]: > typescript
Use the correct capitalization here (TypeScript).

End the sentence with a full stop.

File: source/contributed/typescript-and-screeps-1.md [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+
+## Getting Started with TypeScript
+
+> **Note**: If you already have a Typescript project that compiles, and understand the basic tooling, you can most likely skip the rest of this article
+
+This section will only provide a very brief overview of getting started with TypeScript.  For more in-depth, please see the above Deep Dive link, or the TypeScript docs.
```
Comment (PostCrafter) [inline source/contributed/typescript-and-screeps-1.md]: Maybe add a link to the docs?

File: source/contributed/typescript-and-screeps-1.md [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+
+### Install Your Tooling
+
+To get off the ground with typescript, you're going to need quite a few different tools:
```
Comment (PostCrafter) [inline source/contributed/typescript-and-screeps-1.md]: > typescript
Capitalization

File: source/contributed/typescript-and-screeps-1.md [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+
+* An IDE that supports TS ([Slant](https://www.slant.co/topics/5815/~ides-for-typescript-development) recommends Visual Studio Code as the most popular choice)
+
+* NPM (not in the scope of this article)
```
Comment (PostCrafter) [inline source/contributed/typescript-and-screeps-1.md]: npm always refers to itself as `npm`, I think this article should as well.
Consider adding a link to the website.

File: source/contributed/typescript-and-screeps-1.md [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+
+* The typescript compiler:
```
Comment (PostCrafter) [inline source/contributed/typescript-and-screeps-1.md]: > typescript
Capitalization

File: source/contributed/typescript-and-screeps-1.md [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+  ```bash
+  # for cutting edge in development version (currently 2.4.0):
+  $ npm install -g typescript@next
```
Comment (PostCrafter) [inline source/contributed/typescript-and-screeps-1.md]: I personally really dislike global installations as they put a constraint on the host system.

I do realise though that it can be a lot more tedious to install everything locally (especially without covering npm), so this might be okay in this case.
Alternatively you could just provide the links to the tools and let the use install them himself.

File: source/contributed/typescript-and-screeps-1.md [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+
+### Creating Your First New Project
+> **Note**: This is just a brief guide to getting a bare project going. For most users, the **screeps-typescript-starter** below will be a better & more full-feature choice
+
+> **Note**: Your IDE may complete these steps for you when starting a new project... If so, go ahead and skip down to **Typings**
+
+To get a bare project going, we have to first `init` all our tools.  `cd` into your project directory (presumably, an empty folder), and run the following:
+
+```bash
+$ cd my_proj/
```
Comment (PostCrafter) [inline source/contributed/typescript-and-screeps-1.md]: According to the previous sentence we already are in our project directory, so we won't have to change to it again.

File: source/contributed/typescript-and-screeps-1.md [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+    "typings/**/*.d.ts" /* Tell TypeScript about our installed Typings
+  ]
+}
+```
+
+### Typings
+
+The `typings` tool is how we install TypeScript typings for external JS modules (such as those from `npm`).
+
+Lets install the typings for Screeps, Lodash (screeps is using v3.10.1 at the time of this article), and Node (v6.10.3 at the time of this article).  This way the TypeScript compiler can know about these libraries, even though they aren't actually present in your code
```
Comment (PostCrafter) [inline source/contributed/typescript-and-screeps-1.md]: > Lets
Should be "Let's", also end the sentence with a full stop.

File: source/contributed/typescript-and-screeps-1.md [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+/my_proj$ typings install --save --global "dt~lodash@3.10.0"
+```
+
+#### Newbie Gotcha: Ambient Declarations vs. Module Declarations
+
+As you may have guessed, typing is a key feature of TypeScript.  However, a common "gotcha" for beginners is the difference between "ambient" (also called "global") typing declarations, and "module" typing declarations.
+
+Above, we installed the typings as "global".  This means the declarations inside are just assumed to be available everywhere in your project, without having to `import` or `require` them.
+
+For example, `Creep` is ambient, because we can just call `Creep.moveTo(...)` without having to first call `import { Creep } from "some-module";`.
```
Comment (PostCrafter) [inline source/contributed/typescript-and-screeps-1.md]: > we can just call `Creep.moveTo(...)`
This is probably a bad example, as `Creep.moveTo` would be a static function. Maybe use `Game.creeps` instead.

File: source/contributed/typescript-and-screeps-1.md [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+
+Your own code will largely use "module" declarations.  That is, the declaration and type information won't be available in your files, unless you `import` it first.
+
+TypeScript will automatically infer which type you are using in the file.  If your file contains the `export` keyword anywhere, TS will automatically treat the file as a Module.  If not, it will treat the file as ambient (AKA global).
+
+A common mistake is to accidently include `export` somewhere in a file that is meant to be global, and suddenly have typescript complain about being unable to find your definitions because they suddenly switched from "ambient" to "module".
```
Comment (PostCrafter) [inline source/contributed/typescript-and-screeps-1.md]: > typescript
Capitalization

File: source/contributed/typescript-and-screeps-1.md [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+
+So you have some code.  Now what?
+
+There are a lot of ways to compile typescript for export (the built in compiler, grunt, gulp, webpack, etc, etc).  Rather than filling this guide with more TS-specific information, this section will merely serve as a guide to point you in the right direction.  For a full-featured setup, see the **screeps-typescript-starter** below.
+
+To start with, at a minimum, you will need [webpack](https://webpack.js.org/guides/webpack-and-typescript/) and either [ts-loader](https://github.com/TypeStrong/ts-loader/) or [awesome-typescript-loader](https://github.com/s-panferov/awesome-typescript-loader)
+(loaders integrate typescript with webpack).  This guide recommends **awesome-typescript-loader**:
+
+```bash
+/my_proj$ npm install --save-dev webpack awesome-typescript-loader
```
Comment (PostCrafter) [inline source/contributed/typescript-and-screeps-1.md]: Seems weird to me that you install webpack again, after you installed it globally before.

File: source/contributed/typescript-and-screeps-1.md [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+module.exports = {
+  entry: './src/main.ts',       // the main exported module of your application
+  output: {
+    filename: './main.js',      // the name of the compiled output file
+    libraryTarget: 'commonjs2', // required to export as module
+  },
+
+  target: 'node',               // required for Screeps
+
+  node: {                       // Screeps specific Node Settings
```
Comment (PostCrafter) [inline source/contributed/typescript-and-screeps-1.md]: Promise is also not available, therefore it should be disabled.

File: source/contributed/typescript-and-screeps-2.md [hunk 14]
```diff
@@ -0,0 +1,598 @@
...
+In Part Two, we are going to take a look at using custom TypeScript interfaces and types to extend the Screeps classes to write shorter, more error-proof code.
+Towards the end of the article, we will look at some specific Screeps use-cases that take advantage of these custom types.
+
+## Advanced Typescript & Screeps Typing
+> AKA, how to supercharge your code
+
+### Extending the Default Screeps Typings with declaration merging
+
+TypeScript provides support for [declaration merging](https://www.typescriptlang.org/docs/handbook/declaration-merging.html).  This means that declarations (most commonly, interfaces) defined throughout your code with the same name, will be merged into one large interface.  To use an example from the docs:
+
```
Comment (PostCrafter) [inline source/contributed/typescript-and-screeps-2.md]: I think you should use a screeps related example here, otherwise it will only duplicate the TypeScript documentation.

File: source/contributed/typescript-and-screeps-2.md [hunk 15]
```diff
@@ -0,0 +1,598 @@
...
+// individual memory interfaces that we will extend later in our codebase
+interface CreepMemory {}
+interface FlagMemory {}
+interface RoomMemory {}
+interface SpawnMemory {}
+```
+
+Now we can start actually telling TypeScript about all the information we are storing in our Creeps Memory.
+
+> **Note** tslint will complain about having an empty interface here, as usually that is a symptom of some other problem.  However, in our case it's nice to help keep track of which object we are working with.
```
Comment (PostCrafter) [inline source/contributed/typescript-and-screeps-2.md]: This depends entirely on your setup, some might not use tslint or simply run with a different configuration.

File: source/contributed/typescript-and-screeps-2.md [hunk 16]
```diff
@@ -0,0 +1,598 @@
...
+```
+However, when we get the object back, we lost it's type.  It could be anything.  In the case of `Miner` above, *we know* it's a Source — but the compiler doesn't.
+
+We can tell it, though.  The `getObjectById` method is actually typed as a generic, so we can use:
+```typescript
+// `myObject` will be type of Source
+const myObject = Game.getObjectById<Source>(objectID);
+```
+
+This works good enough for that specific example.  But we want MOAR.
```
Comment (PostCrafter) [inline source/contributed/typescript-and-screeps-2.md]: I don't think we actually want "MOAR". As funny as it sounds, in my opinion the article should be informative.

**Comment by bryanbecker** [comment]
First off, thanks for the feedback.  I'll work on pushing these updates in the next day or so

> it just explains TypeScript features in a screeps environment.

Isn't any tutorial "Just explain X features in a Y environment"?

Look at existing articles:

- Modifying Prototypes (i.e. JS 101 in a Screeps environment)... starts right off with "What are prototypes?"  --- By the way, I think articles like this are a *good* thing
- Advanced Grunt Usage... besides the specific example of using Grunt to upload, the rest is general Grunt tips --- (again, I think this is a *good* thing)

----

One of the 3 content guidelines was
> Completeness: If you write a guide or a how-to, it should leave readers with a finished, working configuration and give them an idea of where to go from there.


@PostCrafter I don't mean this as an attack on your criticism at all (hard to convey conversation tone through messaging), but maybe the guidelines should be revisited first?

Does the Screeps team want: Articles that say "Put this in your code to achieve X.  This is called "whatever".  See <link> for more info", or do we "Here's a new tool, here's how to use it, and here's how to directly apply it to Screeps?"

File: source/contributed/typescript-and-screeps-2.md (hunk 15 above)
Comment (bryanbecker) [inline source/contributed/typescript-and-screeps-2.md]: How about reword this to "tslint (with default settings)"?

File: source/contributed/typescript-and-screeps-2.md (hunk 14 above)
Comment (bryanbecker) [inline source/contributed/typescript-and-screeps-2.md]: Argh... I originally had an example using Memory; must have cut it

File: source/contributed/typescript-and-screeps-1.md (hunk 12 above)
Comment (bryanbecker) [inline source/contributed/typescript-and-screeps-1.md]: nice catch

File: source/contributed/typescript-and-screeps-1.md (hunk 7 above)
Comment (bryanbecker) [inline source/contributed/typescript-and-screeps-1.md]: Agreed, I'll change this to be more clear

File: source/contributed/typescript-and-screeps-1.md (hunk 8 above)
Comment (bryanbecker) [inline source/contributed/typescript-and-screeps-1.md]: I'll debate that generally projects include the commands they just told you to run

That said, I agree it would read better as
> `cd` into your project directory (presumably, an empty folder), and initialize all your project tools:

File: source/contributed/typescript-and-screeps-1.md (hunk 1 above)
Comment (bryanbecker) [inline source/contributed/typescript-and-screeps-1.md]: I was trying to be cheeky... I guess it ended up being confusing... will clarify

**Comment by artch** [comment]
> Does the Screeps team want: Articles that say "Put this in your code to achieve X. This is called "whatever". See linked__article for more info", or do we "Here's a new tool, here's how to use it, and here's how to directly apply it to Screeps?"

Ideally we'd want to have both kinds of articles. So that a complete novice would be able to start quickly from scratch and continue on his own without a lot of reading; or opt-in to discover more advanced explanations and tricks. I'd suggest adding some "Quick Start" section in the beginning.

**Comment by artch** [comment]
So what's the status on this?
